File: apps/server/agentcore/api/routes/conversations/handoff.py
"""Local→云 handoff: snapshot / dispatch a cloud job / list / diff / apply (P2e)."""

import asyncio

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from agentcore.api.dependencies import (
    AuthUser,
    get_conversation_repo,
    get_cost_event_repo,
    get_db,
    get_handoff_job_repo,
)
from agentcore.api.schemas import (
    ApplyHandoffRequest,
    DispatchHandoffRequest,
    HandoffDiffResponse,
    HandoffFileChange,
    HandoffJobListResponse,
    HandoffJobSummary,
)
from agentcore.api.sse import sse_response
from agentcore.conversation.common import resolve_local_binding
from agentcore.conversation.quota import QuotaLimits, enforce_quota
from agentcore.conversation.rate_limit import enforce_user_message_rate_limit
from agentcore.conversation.service import dispatch_handoff
from agentcore.core.error_codes import ErrorCode
from agentcore.core.errors import ConflictError, NotFoundError, ValidationError
from agentcore.core.logging import get_logger
from agentcore.db.repositories import (
    ConversationRepository,
    CostEventRepository,
    HandoffJobRepository,
)
from agentcore.runtime.events import (
    EventSink,
    error_event,
    handoff_apply_done,
    handoff_snapshot_done,
)
from agentcore.storage import SnapshotNotFound
from agentcore.workspace.handoff import snapshot_local
from agentcore.workspace.handoff_apply import ApplySelection, apply_handoff
from agentcore.workspace.handoff_diff import compute_handoff_diff
from agentcore.workspace.handoff_reclaim import reclaim_after_apply
from agentcore.workspace.locate import LocalBinding, build_workspace

from ._helpers import _get_owned_conversation, _require_owned_conversation

router = APIRouter(prefix="/conversations", tags=["conversations"])

logger = get_logger(__name__)

# Terminal statuses where Diff may still be read while snapshots remain.
_DIFFABLE = frozenset({"succeeded", "applied", "discarded"})
# Open for apply / discard of a successful result.
_APPLYABLE = frozenset({"succeeded"})
# Open for discard (completed run, result wanted or not).
_DISCARDABLE = frozenset({"succeeded", "failed"})


async def _run_handoff(
    *,
    user_id: str,
    folder_id: str | None,
    conversation_id: str,
    binding: LocalBinding,
    sink: EventSink,
) -> None:
    """Drive a local→云 handoff snapshot to completion over its SSE sink (P2e / e1).

    Mirrors ``stream_chat``'s shape: the ARCHIVE op is emitted on ``sink`` (the
    bound desktop fulfils it via the ops resolve endpoint), and on success a
    ``handoff_snapshot_done`` carrying the new snapshot id is emitted before the
    stream closes. Any failure is surfaced as an inline ``error`` event (never an
    unhandled crash on this detached task), so the client always learns the outcome.
    """
    try:
        ref = await snapshot_local(
            user_id=user_id,
            folder_id=folder_id,
            conversation_id=conversation_id,
            binding=binding,
            sink=sink,
        )
        sink.emit(
            handoff_snapshot_done(
                snapshot_id=ref.snapshot_id,
                conversation_id=conversation_id,
                size_bytes=ref.size_bytes,
            )
        )
    except Exception as e:
        logger.warning("handoff.failed", conversation_id=conversation_id, error=str(e))
        sink.emit(error_event(ErrorCode.HANDOFF_FAILED, str(e)))
    finally:
        if not sink._closed:
            sink.close(reason="handoff_archive_finally")


@router.post("/{conversation_id}/workspace/handoff")
async def handoff_local_workspace(
    conversation_id: str,
    user: AuthUser,
    conv_repo: ConversationRepository = Depends(get_conversation_repo),
    session: AsyncSession = Depends(get_db),
):
    """Snapshot a local-mode workspace to the cloud over the channel (双模式工作区 P2e / e1).

    Local-mode files live on the user's machine, so the post-turn OSS backup skips
    them; this is the explicit, on-demand 本地→云 snapshot (§四). Streams SSE: a
    ``workspace_op_required`` (the ARCHIVE op) the bound desktop fulfils, then a
    ``handoff_snapshot_done`` carrying the new snapshot id (it lands in the same
    snapshot list / restore / download as cloud-mode versions). 422 when the
    conversation is not in local mode — a cloud workspace already snapshots itself,
    and there is nothing on the user's disk to fetch.

    Binding resolution matches turn routing (``resolve_local_binding``): project
    chats inherit the folder bind; 裸聊 uses ``local_root_id`` or container root.
    """
    conv = await _get_owned_conversation(conversation_id, user.user_id, conv_repo)
    binding = await resolve_local_binding(session, conv)
    if binding is None:
        raise ValidationError("该对话不是本地模式")

    sink = EventSink()
    task = asyncio.create_task(
        _run_handoff(
            user_id=user.user_id,
            # Handoff base snapshots are keyed by source conversation (not the
            # project folder), so they never collide with cloud project snapshots.
            folder_id=None,
            conversation_id=conversation_id,
            binding=binding,
            sink=sink,
        )
    )
    return sse_response(sink, producer=task)


@router.post("/{conversation_id}/workspace/handoff/dispatch")
async def dispatch_handoff_job(
    conversation_id: str,
    body: DispatchHandoffRequest,
    user: AuthUser,
    conv_repo: ConversationRepository = Depends(get_conversation_repo),
    cost_repo: CostEventRepository = Depends(get_cost_event_repo),
    session: AsyncSession = Depends(get_db),
):
    """Hand a task off to a cloud team seeded from the local workspace (P2e / e2).

    The 本地→云 交接 (§四): snapshot the user's local files, then run an Agent team on
    that snapshot in the cloud — autonomously, in the background — so a parallel team
    is not bottlenecked by the single desktop channel. Streams SSE: a
    ``workspace_op_required`` (the ARCHIVE op) the bound desktop fulfils, then a
    ``handoff_job_started`` carrying the job id; the cloud run continues detached
    after the stream closes (poll ``GET …/handoff/jobs`` for its status). 422 when
    the conversation is not in local mode (nothing local to hand off).

    Binding resolution matches turn routing (``resolve_local_binding``).
    Gated like ``send_message`` (it spends tokens): rate limit → ownership → quota.
    """
    await enforce_user_message_rate_limit(user.user_id)
    conv = await _get_owned_conversation(conversation_id, user.user_id, conv_repo)
    await enforce_quota(cost_repo, user.user_id, limits=QuotaLimits.for_user(user))

    binding = await resolve_local_binding(session, conv)
    if binding is None:
        raise ValidationError("该对话不是本地模式")

    sink = EventSink()
    task = asyncio.create_task(
        dispatch_handoff(
            conversation_id=conversation_id,
            user_id=user.user_id,
            # Base snapshot keyed by source conversation (see handoff snapshot).
            folder_id=None,
            binding=binding,
            task=body.task,
            sink=sink,
        )
    )
    return sse_response(sink, producer=task)


@router.get("/{conversation_id}/handoff/jobs", response_model=HandoffJobListResponse)
async def list_handoff_jobs(
    conversation_id: str,
    user: AuthUser,
    conv_repo: ConversationRepository = Depends(get_conversation_repo),
    job_repo: HandoffJobRepository = Depends(get_handoff_job_repo),
):
    """A conversation's local→云 handoff jobs, newest first (双模式工作区 P2e / e2).

    Backs the client's job badge / PR list: poll this to learn when a dispatched
    cloud run finishes (status succeeded / failed). 404 if the conversation is not
    owned.
    """
    await _require_owned_conversation(conversation_id, user.user_id, conv_repo)
    jobs = await job_repo.list_for_source(conversation_id, user_id=user.user_id)
    data = [HandoffJobSummary.model_validate(j) for j in jobs]
    return HandoffJobListResponse(data=data, total=len(data))


@router.get("/{conversation_id}/handoff/jobs/{job_id}", response_model=HandoffJobSummary)
async def get_handoff_job(
    conversation_id: str,
    job_id: str,
    user: AuthUser,
    conv_repo: ConversationRepository = Depends(get_conversation_repo),
    job_repo: HandoffJobRepository = Depends(get_handoff_job_repo),
):
    """One handoff job's status + result snapshots (双模式工作区 P2e / e2).

    404 if the conversation is not owned, or the job is unknown / belongs to a
    different source conversation.
    """
    await _require_owned_conversation(conversation_id, user.user_id, conv_repo)
    job = await job_repo.get_by_id(job_id, user_id=user.user_id)
    if job is None or job.source_conversation_id != conversation_id:
        raise NotFoundError("交接任务不存在")
    return HandoffJobSummary.model_validate(job)


@router.get(
    "/{conversation_id}/handoff/jobs/{job_id}/diff",
    response_model=HandoffDiffResponse,
)
async def get_handoff_job_diff(
    conversation_id: str,
    job_id: str,
    user: AuthUser,
    conv_repo: ConversationRepository = Depends(get_conversation_repo),
    job_repo: HandoffJobRepository = Depends(get_handoff_job_repo),
):
    """A finished handoff's result diff — the change set for the local apply (P2e / e3).

    Compares the team's result snapshot against the base it ran on and returns the
    per-file change set (added / modified / deleted) the desktop replays onto the
    user's local files; each entry carries the base hash for the client's three-way
    conflict check (clean / already-applied / conflict). 404 if the conversation is
    not owned or the job is unknown / from another source conversation; 409 while the
    job has not succeeded yet (no result to diff).
    """
    await _require_owned_conversation(conversation_id, user.user_id, conv_repo)
    job = await job_repo.get_by_id(job_id, user_id=user.user_id)
    if job is None or job.source_conversation_id != conversation_id:
        raise NotFoundError("交接任务不存在")
    if job.status not in _DIFFABLE or not job.result_snapshot_id:
        raise ConflictError("交接任务尚未产出结果")
    try:
        changes = await compute_handoff_diff(
            user_id=user.user_id,
            source_folder_id=None,
            source_conversation_id=conversation_id,
            base_snapshot_id=job.base_snapshot_id,
            job_conversation_id=job.job_conversation_id,
            result_snapshot_id=job.result_snapshot_id,
        )
    except SnapshotNotFound as e:
        raise NotFoundError("交接快照不存在") from e
    data = [HandoffFileChange.model_validate(c) for c in changes]
    return HandoffDiffResponse(
        job_id=job.id,
        data=data,
        total=len(data),
        added=sum(1 for c in data if c.change_type == "added"),
        modified=sum(1 for c in data if c.change_type == "modified"),
        deleted=sum(1 for c in data if c.change_type == "deleted"),
    )


async def _run_apply(
    *,
    user_id: str,
    source_folder_id: str | None,
    source_conversation_id: str,
    job_id: str,
    job_conversation_id: str,
    base_snapshot_id: str,
    result_snapshot_id: str,
    binding: LocalBinding,
    selections: list[ApplySelection],
    sink: EventSink,
) -> None:
    """Drive a handoff result apply to completion over its SSE sink (P2e / e3).

    Builds the desktop-backed ``LocalWorkspace`` over this stream, then replays the
    accepted changes onto the user's machine (WRITE_BYTES / DELETE the bound desktop
    fulfils). On success a ``handoff_apply_done`` carrying the per-file outcomes is
    emitted before the stream closes; a missing snapshot or any failure surfaces as
    an inline ``error`` event (never an unhandled crash on this detached task).

    After a successful apply pass the job is marked ``applied`` and the cloud host
    is soft-deleted (§7.6 结束可收) — soft-delete only; hard purge waits retention.
    """
    try:
        backend = build_workspace(
            user_id=user_id,
            folder_id=source_folder_id,
            conversation_id=source_conversation_id,
            sink=sink,
            local_binding=binding,
        )
        outcomes = await apply_handoff(
            backend=backend,
            user_id=user_id,
            source_folder_id=source_folder_id,
            source_conversation_id=source_conversation_id,
            base_snapshot_id=base_snapshot_id,
            job_conversation_id=job_conversation_id,
            result_snapshot_id=result_snapshot_id,
            selections=selections,
        )
        try:
            await reclaim_after_apply(
                job_id=job_id,
                user_id=user_id,
                job_conversation_id=job_conversation_id,
            )
        except Exception as e:
            # Apply already landed locally — reclaim is best-effort; do not fail the
            # SSE done event (client already has the merge). Retention will age later.
            logger.warning(
                "handoff.reclaim_after_apply_failed",
                job_id=job_id,
                error=str(e),
            )
        sink.emit(
            handoff_apply_done(
                job_id=job_id,
                conversation_id=source_conversation_id,
                results=[o.to_dict() for o in outcomes],
            )
        )
    except SnapshotNotFound as e:
        logger.warning(
            "handoff.apply_snapshot_missing",
            conversation_id=source_conversation_id,
            error=str(e),
        )
        sink.emit(error_event(ErrorCode.HANDOFF_SNAPSHOT_NOT_FOUND, str(e)))
    except Exception as e:
        logger.warning(
            "handoff.apply_failed",
            conversation_id=source_conversation_id,
            error=str(e),
        )
        sink.emit(error_event(ErrorCode.HANDOFF_APPLY_FAILED, str(e)))
    finally:
        if not sink._closed:
            sink.close(reason="handoff_apply_finally")


@router.post("/{conversation_id}/handoff/jobs/{job_id}/apply")
async def apply_handoff_job(
    conversation_id: str,
    job_id: str,
    body: ApplyHandoffRequest,
    user: AuthUser,
    conv_repo: ConversationRepository = Depends(get_conversation_repo),
    job_repo: HandoffJobRepository = Depends(get_handoff_job_repo),
    session: AsyncSession = Depends(get_db),
):
    """Apply a finished handoff's selected changes back to the local workspace (P2e / e3).

    The last leg of the 本地→云 round trip: the user's per-file decisions (take cloud /
    keep local, with the locally-observed hash) are replayed onto their machine over
    the channel. Streams SSE: ``workspace_op_required`` (WRITE_BYTES / DELETE) the
    bound desktop fulfils, then a ``handoff_apply_done`` with the per-file outcomes.
    The conflict gate is server-authoritative — a file that diverged locally since the
    base is refused (status ``conflict``) unless its selection ``force``\\s it.

    On a successful apply pass the job becomes ``applied`` and the cloud host enters
    soft-delete reclaim (§7.6) — not an immediate destroy.

    Binding resolution matches turn routing (``resolve_local_binding``).
    404 if the conversation is not owned or the job is unknown / from another source;
    409 while the job has not succeeded yet, or after applied/discarded; 422 when the
    conversation is not in local mode (nothing local to apply onto).
    """
    conv = await _get_owned_conversation(conversation_id, user.user_id, conv_repo)
    job = await job_repo.get_by_id(job_id, user_id=user.user_id)
    if job is None or job.source_conversation_id != conversation_id:
        raise NotFoundError("交接任务不存在")
    if job.status == "applied":
        raise ConflictError("交接任务已合回")
    if job.status == "discarded":
        raise ConflictError("交接任务已丢弃")
    if job.status not in _APPLYABLE or not job.result_snapshot_id:
        raise ConflictError("交接任务尚未产出结果")

    binding = await resolve_local_binding(session, conv)
    if binding is None:
        raise ValidationError("该对话不是本地模式")

    selections = [
        ApplySelection(path=s.path, decision=s.decision, local_sha=s.local_sha, force=s.force)
        for s in body.selections
    ]
    sink = EventSink()
    task = asyncio.create_task(
        _run_apply(
            user_id=user.user_id,
            source_folder_id=None,
            source_conversation_id=conversation_id,
            job_id=job.id,
            job_conversation_id=job.job_conversation_id,
            base_snapshot_id=job.base_snapshot_id,
            result_snapshot_id=job.result_snapshot_id,
            binding=binding,
            selections=selections,
            sink=sink,
        )
    )
    return sse_response(sink, producer=task)


@router.post(
    "/{conversation_id}/handoff/jobs/{job_id}/discard",
    response_model=HandoffJobSummary,
)
async def discard_handoff_job(
    conversation_id: str,
    job_id: str,
    user: AuthUser,
    conv_repo: ConversationRepository = Depends(get_conversation_repo),
    job_repo: HandoffJobRepository = Depends(get_handoff_job_repo),
):
    """Abandon a finished handoff's cloud replica without applying (§7.6 结束可收).

    Marks the job ``discarded`` and soft-deletes the hidden job host so retention
    can reclaim it. Succeeded (未合回) and failed jobs may be discarded; already
    discarded is idempotent. 409 if the job is still running / pending, or already
    applied. Diff may still work until retention hard-purges snapshots.

    Uses the request-scoped repos (not a detached session) so the status write and
    host soft-delete share the same DB binding as list/get.
    """
    await _require_owned_conversation(conversation_id, user.user_id, conv_repo)
    job = await job_repo.get_by_id(job_id, user_id=user.user_id)
    if job is None or job.source_conversation_id != conversation_id:
        raise NotFoundError("交接任务不存在")
    if job.status == "applied":
        raise ConflictError("交接任务已合回，无法丢弃")
    if job.status == "discarded":
        return HandoffJobSummary.model_validate(job)
    if job.status not in _DISCARDABLE:
        raise ConflictError("交接任务尚未结束，无法丢弃")

    host_id = job.job_conversation_id
    ok = await job_repo.mark_discarded(job.id)
    if not ok:
        # Race: another request applied/discarded between the check and update.
        job = await job_repo.get_by_id(job_id, user_id=user.user_id)
        if job is None or job.source_conversation_id != conversation_id:
            raise NotFoundError("交接任务不存在")
        if job.status == "applied":
            raise ConflictError("交接任务已合回，无法丢弃")
        if job.status == "discarded":
            return HandoffJobSummary.model_validate(job)
        raise ConflictError("交接任务尚未结束，无法丢弃")

    # Best-effort: host may already be gone / never persisted in thin seeds.
    await conv_repo.soft_delete(host_id, user_id=user.user_id)

    job = await job_repo.get_by_id(job_id, user_id=user.user_id)
    if job is None:
        raise NotFoundError("交接任务不存在")
    return HandoffJobSummary.model_validate(job)

File: apps/desktop/src/renderer/services/handoff.ts
import { StreamError } from "@/lib/errors";
import {
  type ChangeType,
  type HandoffApplySelection,
  type HandoffFileChange,
  sha256HexFromBase64,
} from "@/lib/handoff-review";
import {
  BASE_URL,
  api,
  getCsrfHeaders,
  notifyUnauthorized,
  tryRefresh,
} from "@/services/api";
import { performWorkspaceOp } from "@/services/workspaceOps";
import type { components } from "@/types/api.generated";
import type {
  HandoffApplyDonePayload,
  HandoffJobStartedPayload,
  HandoffSnapshotDonePayload,
  SSEEvent,
  WorkspaceOpRequiredPayload,
} from "@/types/events";
import type { WorkspaceOpName } from "@shared/ipc-contract";

type Schemas = components["schemas"];

export interface HandoffResult {
  snapshotId: string;
  sizeBytes: number;
}

/** 派发成功的回执：作业 id 与承载团队回放的隐藏作业对话 id。 */
export interface HandoffJobStarted {
  jobId: string;
  jobConversationId: string;
}

/** 一个本地→云交接作业（双模式工作区 P2e / e2，camelCase 域模型）。 */
export interface HandoffJob {
  id: string;
  sourceConversationId: string;
  jobConversationId: string;
  baseSnapshotId: string;
  resultSnapshotId: string | null;
  task: string;
  status: "pending" | "running" | "succeeded" | "failed";
  error: string | null;
  createdAt: string;
  updatedAt: string;
  finishedAt: string | null;
}

/** 一次交接结果的 diff：变更集 + 表头计数（双模式工作区 P2e / e3）。 */
export interface HandoffDiff {
  jobId: string;
  changes: HandoffFileChange[];
  total: number;
  added: number;
  modified: number;
  deleted: number;
}

/** 应用回写中单文件的结果，对齐后端 `ApplyOutcome`。 */
export interface HandoffApplyResultRow {
  path: string;
  status: "applied" | "skipped" | "conflict" | "error";
  changeType: ChangeType | null;
  detail: string;
}

/** 一次应用回写的汇总：逐文件结果 + 卷起的计数（双模式工作区 P2e / e3）。 */
export interface HandoffApplySummary {
  jobId: string;
  results: HandoffApplyResultRow[];
  applied: number;
  skipped: number;
  conflicts: number;
  errors: number;
}

/** Server handoff-job payload (`/handoff/jobs`), generated from OpenAPI. */
type BackendJob = Schemas["HandoffJobSummary"];

/** Server diff file-change row (`/handoff/jobs/{id}/diff`), generated from OpenAPI. */
type BackendFileChange = Schemas["HandoffFileChange"];

/** Server diff payload (`/handoff/jobs/{id}/diff`), generated from OpenAPI. */
type BackendDiff = Schemas["HandoffDiffResponse"];

function toJob(b: BackendJob): HandoffJob {
  return {
    id: b.id,
    sourceConversationId: b.source_conversation_id,
    jobConversationId: b.job_conversation_id,
    baseSnapshotId: b.base_snapshot_id,
    resultSnapshotId: b.result_snapshot_id,
    task: b.task,
    status: b.status,
    error: b.error,
    createdAt: b.created_at,
    updatedAt: b.updated_at,
    finishedAt: b.finished_at,
  };
}

function toChange(b: BackendFileChange): HandoffFileChange {
  return {
    path: b.path,
    changeType: b.change_type,
    baseSha: b.base_sha,
    resultSha: b.result_sha,
    isBinary: b.is_binary,
    content: b.content,
    sizeBytes: b.size_bytes,
  };
}

/**
 * 一条工作区交接 SSE 流的通用消费器（e1 快照 / e2 派发 / e3 应用共用）。
 *
 * 三条流形状一致：服务端经通道下发若干 `workspace_op_required`（ARCHIVE / WRITE_BYTES
 * / DELETE），本端用既有 `performWorkspaceOp` 在绑定根上履行并回填；末尾发一条「完成」
 * 事件携结果。`onEvent` 只负责认出并映射那条完成事件（返回非 undefined 即为最终结果）。
 *
 * 复用 send 路径的鉴权：access token 过期则刷新一次重放，否则跳登录。这是独立于聊天流
 * 的专用消费器（聊天流改 store，交接流返回值），故不复用 `dispatchSSEEvent`。失败（内联
 * error 事件 / 传输失败 / 流结束仍无结果）抛出，以便 UI 收口。
 */
async function consumeWorkspaceStream<T>(
  path: string,
  conversationId: string,
  opts: { body?: unknown; signal?: AbortSignal },
  onEvent: (event: SSEEvent) => T | undefined,
): Promise<T> {
  const hasBody = opts.body !== undefined;
  const doFetch = () =>
    fetch(`${BASE_URL}${path}`, {
      method: "POST",
      credentials: "include",
      headers: {
        Accept: "text/event-stream",
        ...getCsrfHeaders("POST"),
        ...(hasBody ? { "Content-Type": "application/json" } : {}),
      },
      body: hasBody ? JSON.stringify(opts.body) : undefined,
      signal: opts.signal,
    });

  let response: Response;
  try {
    response = await doFetch();
    if (response.status === 401) {
      const outcome = await tryRefresh();
      if (outcome === "renewed") {
        response = await doFetch();
      } else if (outcome === "auth_dead") {
        notifyUnauthorized();
        throw new StreamError("auth");
      } else {
        throw new StreamError("network");
      }
      if (response.status === 401) {
        notifyUnauthorized();
        throw new StreamError("auth");
      }
    }
  } catch (err) {
    if (err instanceof StreamError) throw err;
    if (err instanceof DOMException && err.name === "AbortError") throw err;
    throw new StreamError("network");
  }
  if (!response.ok) throw new StreamError("http", response.status);

  const reader = response.body?.getReader();
  if (!reader) throw new StreamError("network");

  const decoder = new TextDecoder();
  let buffer = "";
  let result: T | undefined;
  let failure: string | null = null;

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";
      for (const line of lines) {
        if (!line.startsWith("data: ")) continue;
        let event: SSEEvent;
        try {
          event = JSON.parse(line.slice(6)) as SSEEvent;
        } catch {
          continue; // malformed event — skip
        }
        if (event.type === "workspace_op_required") {
          // Fulfil the op (ARCHIVE / WRITE_BYTES / DELETE) against the bound root,
          // exactly as the chat stream does; it POSTs its result to the ops resolve
          // endpoint, settling the paused server-side op so the flow can proceed.
          void performWorkspaceOp(
            event.payload as WorkspaceOpRequiredPayload,
            conversationId,
          );
        } else if (event.type === "error") {
          failure =
            (event.payload as { message?: string }).message ?? "操作失败";
        } else {
          const mapped = onEvent(event);
          if (mapped !== undefined) result = mapped;
        }
      }
    }
  } catch (err) {
    if (err instanceof DOMException && err.name === "AbortError") throw err;
    throw new StreamError("network");
  }

  if (failure) throw new Error(failure);
  if (result === undefined) throw new StreamError("network");
  return result;
}

/**
 * 本地→云交接（双模式工作区 P2e / e1）：把绑定的本地工作区快照到云端。
 *
 * POST 一个 SSE 端点：服务端下发一个 `workspace_op_required`（ARCHIVE op），本端打包
 * 绑定根并回填；服务端解包暂存、快照入 OSS，末尾发 `handoff_snapshot_done` 带新快照 id。
 */
export async function runHandoff(
  conversationId: string,
  signal?: AbortSignal,
): Promise<HandoffResult> {
  return consumeWorkspaceStream(
    `/v1/conversations/${conversationId}/workspace/handoff`,
    conversationId,
    { signal },
    (event) => {
      if (event.type === "handoff_snapshot_done") {
        const p = event.payload as HandoffSnapshotDonePayload;
        return { snapshotId: p.snapshot_id, sizeBytes: p.size_bytes };
      }
      return undefined;
    },
  );
}

/**
 * 把任务交给云端团队（双模式工作区 P2e / e2）：快照本地文件后在云端后台跑一支 Agent
 * 团队。SSE 先下发 ARCHIVE op（本端履行），末尾发 `handoff_job_started` 带作业 id；云端
 * 运行在流关闭后继续，轮询 `listHandoffJobs` 看状态。仅本地模式可派发（否则 422）。
 */
export async function dispatchHandoffJob(
  conversationId: string,
  task: string,
  signal?: AbortSignal,
): Promise<HandoffJobStarted> {
  return consumeWorkspaceStream(
    `/v1/conversations/${conversationId}/workspace/handoff/dispatch`,
    conversationId,
    { body: { task }, signal },
    (event) => {
      if (event.type === "handoff_job_started") {
        const p = event.payload as HandoffJobStartedPayload;
        return { jobId: p.job_id, jobConversationId: p.job_conversation_id };
      }
      return undefined;
    },
  );
}

/** 一个对话的本地→云交接作业，按时间倒序（双模式工作区 P2e / e2）。 */
export async function listHandoffJobs(
  conversationId: string,
): Promise<HandoffJob[]> {
  const res = await api.get<Schemas["HandoffJobListResponse"]>(
    `/v1/conversations/${conversationId}/handoff/jobs`,
  );
  return res.data.map(toJob);
}

/**
 * 一个已完成交接的结果 diff（双模式工作区 P2e / e3）：result 对 base 快照的变更集，
 * 每条携 base 哈希供客户端三方判定。作业未成功时后端返回 409。
 */
export async function getHandoffDiff(
  conversationId: string,
  jobId: string,
): Promise<HandoffDiff> {
  const res = await api.get<BackendDiff>(
    `/v1/conversations/${conversationId}/handoff/jobs/${jobId}/diff`,
  );
  return {
    jobId: res.job_id,
    changes: res.data.map(toChange),
    total: res.total,
    added: res.added,
    modified: res.modified,
    deleted: res.deleted,
  };
}

/**
 * 应用一个已完成交接的所选变更回本地（双模式工作区 P2e / e3）。SSE 下发 WRITE_BYTES /
 * DELETE op（本端履行），末尾发 `handoff_apply_done` 带逐文件结果。冲突门服务端权威：
 * 本地自基线偏离的文件被拒（status `conflict`），除非该选择 `force`。请求体转回后端的
 * snake_case（`local_sha`）。
 */
export async function applyHandoffJob(
  conversationId: string,
  jobId: string,
  selections: HandoffApplySelection[],
  signal?: AbortSignal,
): Promise<HandoffApplySummary> {
  const body = {
    selections: selections.map((s) => ({
      path: s.path,
      decision: s.decision,
      local_sha: s.localSha,
      force: s.force,
    })),
  };
  return consumeWorkspaceStream(
    `/v1/conversations/${conversationId}/handoff/jobs/${jobId}/apply`,
    conversationId,
    { body, signal },
    (event) => {
      if (event.type === "handoff_apply_done") {
        const p = event.payload as HandoffApplyDonePayload;
        return {
          jobId: p.job_id,
          results: p.results.map((r) => ({
            path: r.path,
            status: r.status,
            changeType: r.change_type,
            detail: r.detail,
          })),
          applied: p.applied,
          skipped: p.skipped,
          conflicts: p.conflicts,
          errors: p.errors,
        } satisfies HandoffApplySummary;
      }
      return undefined;
    },
  );
}

/**
 * 逐文件读本地字节并算 sha256 hex，供 e3 三方判定的「第三方输入」。对每个路径在绑定根上
 * 跑 `read_bytes`（服务端回 base64）→ 解码哈希；文件本地不存在/不可读/非桌面环境一律为
 * null（= 本地无此文件）。并发读取，返回 path→sha|null 映射。
 */
export async function readLocalShas(
  rootId: string,
  paths: string[],
): Promise<Map<string, string | null>> {
  const map = new Map<string, string | null>();
  const fsApi = typeof window !== "undefined" ? window.fsApi : undefined;
  if (!fsApi?.workspaceOp) {
    for (const p of paths) map.set(p, null);
    return map;
  }
  await Promise.all(
    paths.map(async (path) => {
      try {
        const res = await fsApi.workspaceOp(
          rootId,
          "read_bytes" as WorkspaceOpName,
          { path },
        );
        map.set(
          path,
          res.ok && typeof res.value === "string"
            ? await sha256HexFromBase64(res.value)
            : null,
        );
      } catch {
        map.set(path, null);
      }
    }),
  );
  return map;
}
